### ISSUE TITLE
possible to disable split/insertion for window?

### ISSUE THREAD

**Issue by yangkennyk** [koekeishiya/yabai#855]
is it possible to disable splitting for a window? I'd like to keep a "sidebar" window and would like to prevent it from ever being split.

**Comment by yanzhang0219** [comment]
Using yabai rule and signal, please refer to https://github.com/koekeishiya/yabai/wiki/Commands#automation-with-rules-and-signals.

**Comment by yangkennyk** [comment]
> Using yabai rule and signal, please refer to https://github.com/koekeishiya/yabai/wiki/Commands#automation-with-rules-and-signals. 

So I was tinkering with using rules but couldn't quite figure out how to get the window to be within the window tiling tree but not be split. Adding the manage=off rule ends up floating the window.

Can you provide a hint on how I could do it?

**Comment by yanzhang0219** [comment]
Using `jq` along with yabai query to get a specific window you need. E.g., `yabai -m query --windows | jq '.[] | select(.floating==1) | .id'` will output the ids of all the floating windows. 

BTW, I am not quite clear about your goal. You want to make a window with some specific title be floating all the time, right?

**Comment by yangkennyk** [comment]
> Using `jq` (https://stedolan.github.io/jq/manual/) along with yabai query to get a specific window you need. E.g., `yabai -m query --windows | jq '.[] | select(.floating==1) | .id'` will output the ids of all the floating windows.
> 
> BTW, I am not quite clear about your goal. You want to make a window with some specific title be floating all the time, right?

Ah let my try to clarify the goal more.

I do not want the specific window to be "floating". I'd like it to take up space like a regular window being managed but I do not want anything ever inserted into that window causing a split. The problem with using floating for what I'm trying to do is windows end up behind the float.

https://github.com/ianyh/Amethyst has the concept of a "main pane" for their layouts. The main pane is a window that is never moved/resized unless you explicitly resize or move it. So all new windows are inserted into the part of the layout that is not the main pane.

![image](https://user-images.githubusercontent.com/6680040/110218515-6e335d00-7e77-11eb-9a75-5088b1e146f8.png)

**Comment by yanzhang0219** [comment]
You can use the `--insert` action. 

For example:
```
alt + shift - up : yabai -m window --insert north
alt + shift - down : yabai -m window --insert south
alt + shift - left : yabai -m window --insert west
alt + shift - right : yabai -m window --insert east
```
to select the position where a new window will insert. 

A short video below as a demo:

https://user-images.githubusercontent.com/11582667/110220610-9624bf80-7e7b-11eb-8522-d8c52ff76abe.mov

**Comment by yangkennyk** [comment]
> If you want an explicit way to prevent a window from being split automatically, I don't think yabai has one.

yep that's exactly what I was trying to do. I wasn't sure if it was possible to set up a rule that would target the window I want and "re-route" the split to another window.

**Comment by yanzhang0219** [comment]
The idea that I come up with immediately is that:
1. We can use a yabai signal to monitor whether the target window is open and if it is opened, we record its frame position in a temp file. The frame position can be gotten from yabai query. 
2. Then, we use another yabai signal to monitor whether other windows are opened and if so, run a script. In the script, we check whether the target window's frame position is changed. If it is changed, it means the target window is split, and we can use window warp to "re-route" the split to another window.

**Comment by yangkennyk** [comment]
> The idea that I come up with immediately is that:
> 
> 1. We can use a yabai signal to monitor whether the target window is open and if it is opened, we record its frame position in a temp file. The frame position can be gotten from yabai query.
> 2. Then, we use another yabai signal to monitor whether other windows are opened and if so, run a script. In the script, we check whether the target window's frame position is changed. If it is changed, it means the target window is split, and we can use window warp to "re-route" the split to another window.

ah. I wasn't sure if there was a signal/rule that could watch for splits, but given your answer it doesn't seem like it and I'll have to use the temp file work around solution you suggested

**Comment by Bellavene** [comment]
#!/bin/zsh
CURRENT_DISPLAY=$(yabai -m query --windows --window | jq '.display')
WINDOWS_ARRAY=$(yabai -m query --windows --space $(yabai -m query --spaces --space | jq '.index') --display $CURRENT_DISPLAY | jq -r 'map(select(.["is-minimized"]==false and .["is-floating"]==false))')
NUMBER_OF_WINDOWS=$(echo $WINDOWS_ARRAY | jq -r 'length')

case $NUMBER_OF_WINDOWS in
	[0-1])
		# yabai -m config split_ratio 0.65
		yabai -m config split_type auto
		;;
	2)
		# yabai -m config split_ratio 0.5
		;;
	3)
    yabai -m config split_ratio 0.5
    yabai -m space --balance x-axis
    yabai -m config split_type horizontal
		;;
	*)
    yabai -m space --balance x-axis
		;;
esac

**Comment by Bellavene** [comment]
Settings
```
split_ratio 0.30
split_type auto
window_placement first_child
```

Signals
```
yabai -m signal --add event=window_created action='yabai-panes; yabai -m window --warp first; yabai -m window --warp prev; yabai -m space --balance x-axis'
yabai -m signal --add event=window_destroyed action='yabai-panes; yabai -m space --balance x-axis'
yabai -m signal --add event=window_minimized action='yabai-panes; yabai -m space --balance x-axis'
yabai -m signal --add event=window_deminimized action='yabai-panes; yabai -m window --warp first; yabai -m window --warp prev; yabai -m space --balance x-axis'
yabai -m signal --add event=application_hidden action='yabai-panes; yabai -m space --balance x-axis'
yabai -m signal --add event=application_visible action='yabai-panes'
yabai -m signal --add event=space_changed action='yabai-panes'
```

yabai-panes
```
#!/bin/zsh
currentsp=$(yabai -m query --spaces --space | jq '.display')
windowcount=$(yabai -m query --windows --space | jq -r '.[]."is-floating"' | grep false | wc -l)
excludes="Menuwhere\|Color Picker\|xQuartz"
countexcludes=$(yabai -m query --windows --space | jq -r '.[].app' | grep "${excludes}" | wc -l)
hiddenwindows=$(yabai -m query --windows --space | jq -r '.[]."is-visible"' | grep false | wc -l)
windowcount="$((windowcount - hiddenwindows - countexcludes))"

case $windowcount in
		[0-1])
			yabai -m config split_ratio 0.30 split_type vertical ;;
		2)
			yabai -m config split_ratio 0.5 split_type horizontal ;;
		*)
			yabai -m config split_ratio 0.5 split_type horizontal ;;
esac
```

**Comment by Bellavene** [comment]
I believe we can close this issue after this was implemented https://github.com/koekeishiya/yabai/issues/2510
